datasource/src/main/scala/quasar/physical/mongo/Mongo.scala

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright 2014–2018 SlamData Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package quasar.physical.mongo
+
+import slamdata.Predef._
+import quasar.api.resource.{ResourceName, ResourcePath}
+import quasar.connector.{MonadResourceErr, ResourceError}
+import quasar.physical.mongo.MongoResource.{Database, Collection}
+
+import cats.effect.{ConcurrentEffect, IO}
+import cats.syntax.eq._
+import fs2.Stream
+import fs2.concurrent._
+import org.bson.{Document => _, _}
+import org.mongodb.scala._
+
+class Mongo[F[_]: MonadResourceErr : ConcurrentEffect] private[Mongo](client: MongoClient) {
+  import Mongo._
+
+  val F: ConcurrentEffect[F] = ConcurrentEffect[F]
+
+  def databases: Stream[F, Database] =
+    observableAsStream(client.listDatabaseNames).map(Database(_))
+
+  def databaseExists(database: Database): Stream[F, Boolean] =
+    databases.exists(_ === database)
+
+  def collections(database: Database): Stream[F, Collection] = for {
+    dbExists <- databaseExists(database)
+    res <- if (!dbExists) { Stream.empty } else {
+      observableAsStream(client.getDatabase(database.name).listCollectionNames()).map(Collection(database, _))
+    }
+  } yield res
+
+  def collectionExists(collection: Collection): Stream[F, Boolean] =
+    collections(collection.database).exists(_ === collection)
+
+  def findAll(collection: Collection): Stream[F, BsonValue] = for {
```

Comment:
I think a return type of `F[Stream[F, BsonValue]]` would be a better fit for the `Datasource` api. The outer `F[_]` represents the existence check and the inner stream the actual data. Additionally, we have `F[_]: MonadResourceErr`,  so prefer `MonadResourceErr.raiseError(...)`.